### PR TITLE
BUG: MultiIndex.union losing extension array dtype

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This covers some cases only, will work on the remaining ones as a follow up

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -3644,6 +3644,10 @@ def _union(self, other, sort) -> MultiIndex:
             # This is only necessary if both sides have nans or one has dups,
             # fast_unique_multiple is faster
             result = super()._union(other, sort)
+
+            if isinstance(result, MultiIndex):
```
Comment (mroeschke) [inline pandas/core/indexes/multi.py:3648]: Could `self._is_multi` be used?

**Review by mroeschke — COMMENTED** [review]
The other cases that don't work are the non-numeric ones?

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (phofl) [reply pandas/core/indexes/multi.py:3648]: No, we have to check result, but result could also be an array, so have to be explicit here

**Comment by phofl** [comment]
Ah no, sorry. This was not explicit enough. It does not work if we run into the ``algos.union_with_duplicates(lvals, rvals)`` code branch in ``_union()``. This is something for a follow up

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**